File: eco-core/core-plugin/src/main/java/com/willfp/eco/spigot/gui/GUIListener.java
package com.willfp.eco.spigot.gui;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.PluginDependent;
import com.willfp.eco.core.gui.menu.Menu;
import com.willfp.eco.core.gui.slot.Slot;
import com.willfp.eco.internal.gui.menu.EcoMenu;
import com.willfp.eco.internal.gui.menu.MenuHandler;
import com.willfp.eco.internal.gui.slot.EcoSlot;
import org.apache.commons.lang.Validate;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.jetbrains.annotations.NotNull;

public class GUIListener extends PluginDependent<EcoPlugin> implements Listener {
    /**
     * Pass an {@link EcoPlugin} in order to interface with it.
     *
     * @param plugin The plugin to manage.
     */
    public GUIListener(@NotNull final EcoPlugin plugin) {
        super(plugin);
    }

    @EventHandler
    public void handleSlotClick(@NotNull final InventoryClickEvent event) {
        if (!(event.getWhoClicked() instanceof Player)) {
            return;
        }

        if (event.getClickedInventory() == null) {
            return;
        }
        Menu menu = MenuHandler.getMenu(event.getClickedInventory());
        if (menu == null) {
            return;
        }

        int row = Math.floorDiv(event.getSlot(), 9);
        int column = event.getSlot() - (row * 9);

        Slot slot = menu.getSlot(row, column);

        Validate.isTrue(slot instanceof EcoSlot, "Slot not instance of EcoSlot!");

        EcoSlot ecoSlot = (EcoSlot) menu.getSlot(row, column);
        event.setCancelled(true);
        ecoSlot.handleInventoryClick(event);
    }

    @EventHandler
    public void handleClose(@NotNull final InventoryCloseEvent event) {
        if (!(event.getPlayer() instanceof Player)) {
            return;
        }

        Menu menu = MenuHandler.getMenu(event.getInventory());

        Validate.isTrue(menu instanceof EcoMenu, "Menu not instance of EcoMenu!");

        EcoMenu ecoMenu = (EcoMenu) menu;

        ecoMenu.handleClose(event);

        this.getPlugin().getScheduler().run(() -> MenuHandler.unregisterMenu(event.getInventory()));
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/eco/spigot/gui/GUIListener.kt
package com.willfp.eco.spigot.gui

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.PluginDependent
import com.willfp.eco.core.drops.DropQueue
import com.willfp.eco.internal.gui.menu.EcoMenu
import com.willfp.eco.internal.gui.menu.MenuHandler
import com.willfp.eco.internal.gui.slot.EcoSlot
import com.willfp.eco.util.MenuUtils
import org.apache.commons.lang.Validate
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.inventory.ClickType
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent

class GUIListener(plugin: EcoPlugin) : PluginDependent<EcoPlugin>(plugin), Listener {
    @EventHandler
    fun handleSlotClick(event: InventoryClickEvent) {
        val player = event.whoClicked
        if (player !is Player) {
            return
        }
        val menu = MenuHandler.getMenu(event.clickedInventory ?: return) ?: return
        val rowColumn = MenuUtils.convertSlotToRowColumn(event.slot)
        val row = rowColumn.first!!
        val column = rowColumn.second!!
        val slot = menu.getSlot(row, column)
        Validate.isTrue(slot is EcoSlot, "Slot not instance of EcoSlot!")
        val ecoSlot = menu.getSlot(row, column) as EcoSlot
        event.isCancelled = true
        ecoSlot.handleInventoryClick(event, menu)

        plugin.scheduler.run{ MenuHandler.getExtendedInventory(event.clickedInventory!!).refresh(player) }
    }

    @EventHandler
    fun handleCaptivatorSlots(event: InventoryClickEvent) {
        val player = event.whoClicked
        if (player !is Player) {
            return
        }

        val menu = MenuHandler.getMenu(player.openInventory.topInventory) ?: return

        plugin.scheduler.run{ MenuHandler.getExtendedInventory(player.openInventory.topInventory).refresh(player) }
        plugin.scheduler.runLater({ Bukkit.getLogger().info(menu.getCaptiveItems(player).toString()) }, 1)
    }

    @EventHandler
    fun handleShiftClick(event: InventoryClickEvent) {
        if (!(event.click == ClickType.SHIFT_RIGHT || event.click == ClickType.SHIFT_LEFT)) {
            return
        }

        val player = event.whoClicked
        if (player !is Player) {
            return
        }

        val inv = player.openInventory.topInventory

        val menu = MenuHandler.getMenu(inv) ?: return

        val rowColumn = MenuUtils.convertSlotToRowColumn(inv.firstEmpty())
        val row = rowColumn.first!!
        val column = rowColumn.second!!
        val slot = menu.getSlot(row, column)

        if (!slot.isCaptive) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    fun handleClose(event: InventoryCloseEvent) {
        val player = event.player
        if (player !is Player) {
            return
        }
        val menu = MenuHandler.getMenu(event.inventory) ?: return
        Validate.isTrue(menu is EcoMenu, "Menu not instance of EcoMenu!")
        val ecoMenu = menu as EcoMenu
        ecoMenu.handleClose(event)

        DropQueue(player)
            .addItems(ecoMenu.getCaptiveItems(player))
            .setLocation(player.location)
            .forceTelekinesis()
            .push()

        plugin.scheduler.run { MenuHandler.unregisterMenu(event.inventory) }
    }
}